mayastor/src/bdev/dev/nvmx/channel.rs

```diff
@@ -58,13 +61,139 @@ impl<'a> NvmeIoChannel<'a> {
         }
     }
 }
+pub struct IoQpair {
+    qpair: NonNull<spdk_nvme_qpair>,
+    ctrlr_handle: NonNull<spdk_nvme_ctrlr>,
+}
+
+impl IoQpair {
+    fn get_default_options(
+        ctrlr_handle: *mut spdk_nvme_ctrlr,
+    ) -> spdk_nvme_io_qpair_opts {
+        let mut opts = spdk_nvme_io_qpair_opts::default();
+        let default_opts = nvme_bdev_running_config();
+
+        unsafe {
+            spdk_nvme_ctrlr_get_default_io_qpair_opts(
+                ctrlr_handle,
+                &mut opts,
+                size_of::<spdk_nvme_io_qpair_opts>() as u64,
+            )
+        };
+
+        opts.io_queue_requests =
+            max(opts.io_queue_requests, default_opts.io_queue_requests);
+        opts.create_only = true;
+
+        opts
+    }
+
+    /// Create a qpair with default options for target NVMe controller.
+    fn create(
+        ctrlr_handle: *mut spdk_nvme_ctrlr,
+        ctrlr_name: &str,
+    ) -> Result<Self, CoreError> {
+        let qpair_opts = IoQpair::get_default_options(ctrlr_handle);
+
+        let qpair: *mut spdk_nvme_qpair = unsafe {
+            spdk_nvme_ctrlr_alloc_io_qpair(
+                ctrlr_handle,
+                &qpair_opts,
+                size_of::<spdk_nvme_io_qpair_opts>() as u64,
+            )
+        };
+
+        if qpair.is_null() {
+            error!(?ctrlr_name, "Failed to allocate I/O qpair for controller",);
+            Err(CoreError::GetIoChannel {
+                name: ctrlr_name.to_string(),
+            })
+        } else {
+            debug!(?qpair, ?ctrlr_name, "qpair created for controller");
+            Ok(Self {
+                qpair: NonNull::new(qpair).unwrap(),
+                ctrlr_handle: NonNull::new(ctrlr_handle).unwrap(),
```

Comment:
do we need to return an error if ctrlr_handle is null?

Reply:
If it's null, qpair allocation will fail and we end up in 'if qpair.is_null() {' branch.

Reply:
But I can add an assertion against null, as null must not be passed.